jupyterhub/app.py

```diff
@@ -1966,25 +1965,27 @@ async def init_groups(self):
                 group.users.append(user)
         db.commit()
 
-    async def init_roles(self):
+    async def init_role_creation(self):
         """Load default and predefined roles into the database"""
-        db = self.db
-        # tokens are added separately
-        role_bearers = ['users', 'services', 'groups']
-
-        # load default roles
         self.log.debug('Loading default roles to database')
         default_roles = roles.get_default_roles()
         for role in default_roles:
-            roles.create_role(db, role)
+            roles.create_role(self.db, role)
 
+    async def init_role_assignment(self):
+        # tokens are added separately
+        role_bearers = ['users', 'services', 'groups']
+
+        db = self.db
```

Comment:
Let's also implement the admin role assignment here, instead of in `init_users`: query users and services with admin=True, then assign them to the admin role.

Reply:
If I'm not mistaken, that's covered a bit further down, in `roles.check_for_default_roles(db, bearer)`, which calls `roles.assign_default_roles` which queries for the admin flag

Reply:
It should be unconditional, though. Always assigning admin users to the admin role. Right now, that's only done in the `if not admin_role.users` case, which I think doesn't behave correctly (no users with admin role doesn't mean anything is wrong or that a db upgrade has occurred, because having no admins is valid.)

i.e. `user.admin` should always exactly match the admin role assignment, in both directions - no user should have `.admin = True` without being in the admin group, and no user should have the admin role without `.admin = True`.

Reply:
Good point, I'll address it in the next commit